src/gatsby-node.ts

```diff
@@ -106,6 +99,49 @@ export const sourceNodes: NonNullable<GatsbyNode['sourceNodes']> = async (
   writeTypePathsActivity.end()
 }
 
+export const sourceNodes: NonNullable<GatsbyNode['sourceNodes']> = async (
+  gatsbyContext: SourceNodesArgs,
+  pluginOptions: PluginOptions,
+) => {
+  const { reporter, store, webhookBody, getNodes, actions: { touchNode } } = gatsbyContext
+  const { program } = store.getState()
+
+  /**
+   * Validate plugin options. Set default options where necessary. If any
+   * plugin options are invalid, stop immediately.
+   */
+  try {
+    pluginOptions = validatePluginOptions(pluginOptions)
+  } catch (error) {
+    reporter.error(msg('invalid plugin options'))
+    reporter.panic(error)
+  }
+
+  const { typeDefs, typePaths } = schemasToTypeDefs(
+    pluginOptions.schemas,
+    gatsbyContext,
+  )
+
+  if(!webhookBody || JSON.stringify(webhookBody) === "{}") {
+    /** Initial build or rebuild everything */
+    createPrismicTypes(pluginOptions, gatsbyContext, typeDefs)
+    await buildAll(pluginOptions, gatsbyContext, typePaths)
+    writeTypePaths(pluginOptions, gatsbyContext, typePaths, program)
+
+  } else if(isPrismicWebhook(webhookBody) && validateSecret(pluginOptions, webhookBody)) {
+    /** Respond to the webhook here */
+    
+    // touch nodes to prevent garbage collection
+    getNodes().forEach(node => touchNode({ nodeId: node.id}))
+
+    createPrismicTypes(pluginOptions, gatsbyContext, typeDefs)
+
+    const prismicWebhook = webhookBody as PrismicWebhook;
+    await handleWebhook(pluginOptions, gatsbyContext, typePaths, prismicWebhook)
+
+  }
```

Comment:
I was thinking that other webhooks could be sent to other plugins, which would be a case where we don't want to trigger a full rebuild. 

I guess no-one is going to be able to guess the url to send the webhook to so we should be fine :)

I'll make the change :)

Reply:
I'm actually going to get some input on this...I think it needs to change but I'm not convinced where it needs to go...

Reply:
Okay, sorry for the confusion (on my part)...it did need to change, but now we have to add somethings. So If there is no `webhookBody` the plugin should source data. Because you rely on a web-hook body to do inc builds, this means you'll have to do a full data fetch. If there is a web-hook body that comes from Prismic you can to an incremental data fetch. If the web-hook body is not from Prismic, you can safely ignore it and not run any data fetching. HOWEVER, **you still must touch all existing nodes or they will get garbage collected**.

Reply:
Neat :)

I'll get that part done today.